pulsar/src/main/java/com/hazelcast/jet/contrib/pulsar/PulsarSinkBuilder.java

```diff
@@ -0,0 +1,220 @@
+        FunctionEx<? super E, Long> localExtractTimestampFn = extractTimestampFn;
+
+        return SinkBuilder.sinkBuilder("pulsar-sink", ctx -> new PulsarSinkContext(ctx.logger(), localTopic,
+                localConnectionSupplier.get(), localProducerConfig, localSchemaSupplier, localExtractValueFn,
+                localExtractKeyFn, localExtractPropertiesFn, localExtractTimestampFn))
+                .<M>receiveFn(PulsarSinkContext::add)
+                .destroyFn(PulsarSinkContext::destroy)
+                .preferredLocalParallelism(preferredLocalParallelism)
+                .build();
+    }
+
+
+    private static final class PulsarSinkContext<E, M> {
+        private final ILogger logger;
+        private final PulsarClient client;
+        private final Producer<M> producer;
+        private final FunctionEx<? super E, M> extractValueFn;
+        private final FunctionEx<? super E, String> extractKeyFn;
+        private final FunctionEx<? super E, Map<String, String>> extractPropertiesFn;
+        private final FunctionEx<? super E, Long> extractTimestampFn;
+
+        private PulsarSinkContext(
+                @Nonnull ILogger logger,
+                @Nonnull String topic,
+                @Nonnull PulsarClient client,
+                @Nonnull Map<String, Object> producerConfig,
+                @Nonnull SupplierEx<Schema<M>> schemaSupplier,
+                @Nonnull FunctionEx<? super E, M> extractValueFn,
+                @Nullable FunctionEx<? super E, String> extractKeyFn,
+                @Nullable FunctionEx<? super E, Map<String, String>> extractPropertiesFn,
+                @Nullable FunctionEx<? super E, Long> extractTimestampFn
+        ) throws PulsarClientException {
+            this.logger = logger;
+            this.client = client;
+            this.producer = client
+                    .newProducer(schemaSupplier.get())
+                    .topic(topic)
+                    .loadConf(producerConfig)
+                    .create();
+            this.extractKeyFn = extractKeyFn;
+            this.extractValueFn = extractValueFn;
+            this.extractPropertiesFn = extractPropertiesFn;
+            this.extractTimestampFn = extractTimestampFn;
+        }
+
+        private void add(E item) {
+            TypedMessageBuilder<M> messageBuilder = producer.newMessage()
+                                                            .value(extractValueFn.apply(item));
+            if (extractKeyFn != null) {
+                messageBuilder = messageBuilder.key(extractKeyFn.apply(item));
+            }
+            if (extractPropertiesFn != null) {
+                messageBuilder = messageBuilder.properties(extractPropertiesFn.apply(item));
+            }
+            if (extractTimestampFn != null) {
+                messageBuilder.eventTime(extractTimestampFn.apply(item));
+            }
+            messageBuilder.sendAsync();
```

Comment:
This way the sink has no guarantee. If we could ensure that all async calls are flushed, we could provide at-least-once. Also, how do you check for async errors?